Update docs for get_future support

agolynski opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58107 Update docs for get_future support**

Differential Revision: [D28387374](https://our.internmc.facebook.com/intern/diff/D28387374)

agolynski committed:
Update docs for get_future

[ghstack-poisoned]

agolynski committed:
Update on "Update docs for get_future"




[ghstack-poisoned]

agolynski committed:
Update on "Update docs for get_future support"




[ghstack-poisoned]

SciPioneer commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1461,7 +1462,12 @@ Example::
+                tensors are ready.
+                For GPU work, ``fut.done()`` returns true only whether the operation has been enqueued.
+                For mixed CPU-GPU work (e.g. sending GPU tensors with GLOO), ``fut.done()`` returns
+                true when tensors have arrived on respective nodes, but not yet necessarily syched on
```
typo: syched -> synced

SciPioneer commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1445,7 +1445,8 @@ Example::
             .. warning ::
-                ``get_future`` API supports only NCCL backend.
+                ``get_future`` API supports NCCL, and partially GLOO and MPI backends
+                (support is limited to non peer-to-peer operations).
```
Feel it's better to explicitly mention the operators here with some examples.

You can reword it as "no support for peer-to-peer operations like send/recv"

SciPioneer commented on docs/source/distributed.rst:
```diff
@@ -361,6 +361,9 @@ is guaranteed to support two methods:
   output can be utilized on the default stream without further synchronization.
+* ``get_future()`` - returns ``torch._C.Future`` object. Supported for NCCL, also supported for most operations on GLOO
+  and MPI, except for peer to peer operations.
+  Note: as we continue adopting Futures and merging APIs, ``get_future()`` call might become redundant.
```
Is there any relevant GH issue to describe the future merging work?

agolynski replied:
not yet, I should file a small RFC on which functions should be exposed in the resulting merge

SciPioneer commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1461,7 +1462,12 @@ Example::
                 callback and a ``CUDAEvent`` that recorded the callback stream.
 
-                Note that ``fut.done()`` returns only whether the operation has been enqueued on the GPU.
+                For CPU work, ``fut.done()`` returns true when work has been complted and value()
```
Better to itemize these description in the same paragraph:
1) For CPU work, ...
2) For GPU work, ...
3) For mixed CPU-GPU work, ...

When this docstring is included by an rst file and displayed as a webpage, this can give a more clear view.

SciPioneer commented on torch/nn/parallel/distributed.py:
```diff
@@ -1187,7 +1189,8 @@ def register_comm_hook(self, state: object, hook: callable):
 
         .. warning ::
-            ``get_future`` API supports only NCCL backend and will return a ``torch._C.Future``
+            ``get_future`` API supports NCCL, and partially GLOO and MPI backends (support is limited
```
Does it mean that DDP comm hook (assume it only invokes allreduce API and no send/recv) can be supported on all the 3 backends now?

agolynski replied:
yes

agolynski commented:
@agolynski has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387374).

agolynski committed:
Update on "Update docs for get_future support"


Differential Revision: [D28387374](https://our.internmc.facebook.com/intern/diff/D28387374)

[ghstack-poisoned]

agolynski commented:
@agolynski has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387374).
